tempesta_fw/http.c

```diff
@@ -3943,10 +3943,25 @@ tfw_http_msg_process(void *conn, TfwFsmData *data)
 {
 	int r = T_OK;
 	unsigned int trail = data->trail;
-	struct sk_buff *next;
+	struct sk_buff *next, *last_skb;
 
-	if (data->skb->prev)
+	if (data->skb->prev) {
+		last_skb = data->skb->prev;
 		data->skb->prev->next = NULL;
+	} else {
+		last_skb = data->skb;
+		while (last_skb->next != NULL)
+			last_skb = last_skb->next;
+	}
+
+	while (last_skb->len <= trail) {
```

Comment:
OK, I'm good with the loop - the condition in improbable (please mark it as `unlikely` BTW) and the loop is short. We *assume* so. But I don't like the first loop: firstly, it looks like a workaround for a 'bad' API - if we need last item in the list, then we always have to have `skb->prev`; secondly, it's unclear how long the loop could be. And finally I see no reason to traverse the list 3 time while in the first and second loop we have the same termination condition.

Recently a loop was introduced for cache-constructed skbs - it's not good as well, having in mind that skb is in most 64KB and current limit for cached response is IIRC 1GB. However, generally the behaviour is wrong and to be fixed in #498. However, introducing more and more loops on skb list is a bad practice.

Reply:
I've removed the need for loops by passing everywhere a usable data length instead of a tail size. That way we can track whenever we are out of data, so the trailing skb's in a chain can be freed (see `tfw_http_msg_process()`).